### PR TITLE
feat(builder-indexes): cli flag to build Images map 

### PR DESCRIPTION
**Pull request by Lowaiz**
feat(builder-indexes): cli flag to build Images map without scanning docker registries

**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area cmd

/area pkg

/area docs

**What this PR does / why we need it**:

New flag to be able to set the builder.Images map with data given via cli to not search and parse Docker registries.
The flag `builderindex` is string slice, that I parse by splitting with `;` character, considering the first term as the image name, the second term as the target, and everything else as GCC versions available on the image.

I guess the format is not optimum, but I can't find other and simpler way to parse this kind of info from CLI:

Usage: --builderindex "`<image>`;`<target>`;`<gcc-version>`;`<gcc-version>`;`...`" 
Example: --builderindex "`falcosecurity/driverkit-builder`;`any`;`4.9.0`;`5.0.0`;`6.0.0`;`8.0.0`"

Multiple --buildindex can be set to insert different images

**Which issue(s) this PR fixes**:

None, context here: #256

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
- New CLI flag '--builderindex' making able to set up the Images map from CLI.
  Usage: --builderindex "<image>;<target>;<gcc-versuin>;<gcc-versuin>;...".
  Example: --builderindex "falcosecurity/driverkit-builder;any;4.9.0;5.0.0;6.0.0;8.0.0"
```

### CODE REVIEW

**Comment by FedeDP** [comment]
Thanks for this PR!
I was wondering whether to use the same `builderrepo` option instead; like:
* if `builderrepo` points to a local file, we then expect it to provide a list of `"<image>;<target>;<gcc-version>,<gcc-version>,<gcc-version>\n`
* else, we expect it to be a docker repo

It is simpler; what i don't really like about having the new option (thinking twice about it) is how it plays together with `builderrepo` for example, given that they both provide the same informations.

Moreover, it would be huge to abstract everything behind an interface-like object, eg:
```go
type ImagesLister interface {
   ListImage() []Images
}

type FileImagesLister struct {
  *os.File
}

func (f *FileImagesLister) ListImage() []Images { ... }

type RepoImagesLister struct {
  repo string
}

func (f *RepoImagesLister) ListImage() []Images { ... }

RootOptions.go:
for builderrepo := range builderrepos {
  if isFile(builderrepo) {
     builderepostlister.append(newFileImagesLister(builderrepo))
 else
    builderepostlister.append(newRepoImagesLister(builderrepo))
}
```
Then basically, `images.findImage()` will receive the list of `builderepostlister`(that will actually be stored in `builder.Build`), and calls `ListImage()` on each one. 
This way, we also support using `image` subcmd with the new feat.

WDYT?
I know it is a bigger change but it will be much better IMHO!

**Comment by Lowaiz** [comment]
Hey,

That's a clever idea. And when changing to falctoctl later will be simpler.

File: pkg/driverbuilder/builder/image.go
Comment (FedeDP) [inline pkg/driverbuilder/builder/image.go:34]: I'd avoid storing the `Build` here! Instead, i'd pass it to the `ListImages` methods, as a param.
I think it is better for these methods to be stateless: just return a list of `Image`s.
Moreover, given that `regs` are now just used by `RepoImagesLister`, i think they can just be declared inside the `ListImages` method.

File: pkg/driverbuilder/builder/image.go
Comment (FedeDP) [inline pkg/driverbuilder/builder/image.go:39]: Why are you exposing this method now?

File: pkg/driverbuilder/builder/builders.go
Comment (FedeDP) [inline pkg/driverbuilder/builder/builders.go:143]: Why are you exposing this method now?

File: pkg/driverbuilder/builder/image.go
Comment (FedeDP) [reply pkg/driverbuilder/builder/image.go:34]: I think they can also be global const variables in image.go, and then referenced in each `RepoImagesLister`; i don't like adding yet another non-build related field to `Build` struct.
Given it is only used by `RepoImagesLister`, i think it should live there!

File: pkg/driverbuilder/builder/image.go
Comment (FedeDP) [inline pkg/driverbuilder/builder/image.go:87]: I'd just make this method return a slice of Images, it is a much better API (and is unit testable); then main `image.ListImages` will gather all the images and fill the build.Images map.

Comment (Lowaiz) [reply pkg/driverbuilder/builder/image.go]: Found a way to eliminate ref to builder inside objects and function of ImageLister

File: pkg/driverbuilder/builder/image.go
Comment (FedeDP) [inline pkg/driverbuilder/builder/image.go:159]: I think the best here would be to properly capture the type in the regexes, so that we can just do `buildImage.Target = Type(regType)`
be it `any` or a specific target.

Comment (FedeDP) [reply pkg/driverbuilder/builder/image.go:159]: And we can then skip the `target-placeholder` trick below.

File: pkg/driverbuilder/builder/image.go
Comment (FedeDP) [inline pkg/driverbuilder/builder/image.go:119]: I'd keep this check, to make sure that we loaded at least an image.

File: cmd/root_options.go
Comment (FedeDP) [inline cmd/root_options.go:138]: I'd inizialize it as a default value during declaration, using the `default:` tag; we use this library: https://github.com/creasty/defaults#usage

As you can see, a 
```
BuilderRepos []string `default: ["docker.io/falcosecurity/driverkit"] validate:"omitempty" name:"docker repositories to look for builder images or absolute path pointing to a file container builder image index"`
```
Is enough.

**Comment by FedeDP** [comment]
Last round of review, i think after these ones will be addressed, we will be ready :rocket:

**Comment by FedeDP** [comment]
Btw this is already good to go; if you are willing to spend a little bit more time on refining this, it's great! Otherwise i can do that myself on top of yours once it's merged :)
Sorry for the long list of "nit picks", but i feel like we are doing a really good future-proof job here!

File: cmd/root_options.go
Comment (Lowaiz) [reply cmd/root_options.go:138]: Oh I didn't know we could place complex types inside this default statement

**Comment by Lowaiz** [comment]
@FedeDP 
I guess, it's good now !

**Comment by FedeDP** [comment]
Yep @Lowaiz!
Thanks for your patience! It LGTM! :)

File: cmd/root.go
Comment (FedeDP) [reply cmd/root.go:147]: Btw what if we used a structured text (like a yaml) instead of a txt file?
That way we could just parse it to a `[]Image` and be good to go!

Comment (FedeDP) [reply cmd/root.go:147]: (we can do this later before next release!)

Comment (Lowaiz) [reply cmd/root.go:147]: Hi, getting a yaml can be good yes, applying that right now !

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**
